File: scripts/pipeline.R
##### Background #####
# User-facing entry point for retraining the model. Builds on the functions in
# model_training.R and feature_processing.R - source both before this one.


##### Load default data #####

# Loads defaultLabelData.csv and returns the default trainSites list directly (see
# runTrainingPipeline()): every site with response "X1" as positive, every site with
# response "X0" as negative. Feature values for training come from `datFeats` (see
# feature_processing.R's prepareFeatures()), so the two stay consistent.
loadDefaultTrainSites <- function(dataDir = "data"){
  labelData <- read.csv(file.path(dataDir, "defaultLabelData.csv"), row.names = 1)
  list(
    pos = rownames(labelData)[labelData$response == "X1"],
    neg = rownames(labelData)[labelData$response == "X0"]
  )
}


##### Build a training set from datFeats #####

# Builds a response + features data.frame by pulling posSites/negSites rows out of datFeats.
# posSites/negSites: character vectors of rownames(datFeats) - the positive (X1) and
# negative (X0) training sites.
buildTrainingData <- function(datFeats, posSites, negSites){
  overlap <- intersect(posSites, negSites)
  if (length(overlap) > 0) {
    stop("Sites cannot be in both posSites and negSites: ", paste(head(overlap, 5), collapse = ", "))
  }
  missingPos <- setdiff(posSites, rownames(datFeats))
  if (length(missingPos) > 0) {
    warning(length(missingPos), " posSites are not present in datFeats and will be dropped: ",
            paste(head(missingPos, 5), collapse = ", "))
    posSites <- setdiff(posSites, missingPos)
  }
  missingNeg <- setdiff(negSites, rownames(datFeats))
  if (length(missingNeg) > 0) {
    warning(length(missingNeg), " negSites are not present in datFeats and will be dropped: ",
            paste(head(missingNeg, 5), collapse = ", "))
    negSites <- setdiff(negSites, missingNeg)
  }
  response <- factor(rep(c("X1", "X0"), c(length(posSites), length(negSites))), levels = c("X1", "X0"))
  cbind(response = response, datFeats[c(posSites, negSites), ])
}


##### Full training pipeline #####

# datFeats: data.frame of features for all sites (rownames "<uniprot>_<site>"). Final
#   predictions are made for every row of this data.frame.
# trainSites: list(pos = ..., neg = ...) of rowname vectors of datFeats, giving the positive
#   (X1) and negative (X0) training sites. See loadDefaultTrainSites() for the default.
# numReps: number of repeated protein-stratified k-fold train-test splits to generate.
#   Ignored if customSplits is supplied
# k: number of folds per repeat
# customSplits: optional pre-built train-test splits (as produced by makeFoldsProtstrat() /
#   makeRepeatedProtstratSplits()), used instead of auto-generating splits from numReps/k
# splitSeed: base seed used when auto-generating splits
# trainSeed: base seed used for model training
# returnAllPredsLst: if TRUE, also return each split's model's individual predictions on
#   datFeats (allPredsDf is already their per-site median across splits)
#
# Returns a list (as produced by trainAndMakePreds_logreg()):
#   AUCdf            - AUC of every train-test split, on its own held-out test set
#   modsPredsTest_df - the underlying test-set predictions behind AUCdf
#   allPredsDf       - median predicted score for every site in datFeats, across all splits
#   mods             - the trained model for each split
#   allPredsLst      - only present if returnAllPredsLst = TRUE
runTrainingPipeline <- function(datFeats,
                                trainSites,
                                numReps = 1,
                                k = 5,
                                customSplits = NULL,
                                splitSeed = 123,
                                trainSeed = 80,
                                returnAllPredsLst = FALSE){
  trainDat <- buildTrainingData(datFeats = datFeats, posSites = trainSites$pos, negSites = trainSites$neg)

  if (is.null(customSplits)) {
    trainTestSplits <- makeRepeatedProtstratSplits(data = trainDat, numReps = numReps, k = k, seed = splitSeed)
  } else {
    trainTestSplits <- customSplits
  }

  trainAndMakePreds_logreg(
    trainTestSplits = trainTestSplits,
    datFeats = datFeats,
    seed = trainSeed,
    returnAllPredsLst = returnAllPredsLst
  )
}

File: scripts/python/pipeline.py
"""User-facing entry point for retraining the model. Python port of scripts/pipeline.R -
builds on model_training.py (import that module first, e.g. by adding scripts/python to
sys.path).
"""

import os
import warnings

import pandas as pd

from model_training import make_repeated_protstrat_splits, train_and_make_preds_logreg


##### Load default data #####

def load_default_train_sites(data_dir="data"):
    """Loads defaultLabelData.csv and returns the default train_sites dict directly (see
    run_training_pipeline()): every site with response "X1" as positive, every site with
    response "X0" as negative. Feature values for training come from `dat_feats` (see
    feature_processing.py's prepare_features()), so the two stay consistent.
    """
    label_data = pd.read_csv(os.path.join(data_dir, "defaultLabelData.csv"), index_col=0)
    return {
        "pos": label_data.index[label_data["response"] == "X1"].tolist(),
        "neg": label_data.index[label_data["response"] == "X0"].tolist(),
    }


##### Build a training set from dat_feats #####

def build_training_data(dat_feats, pos_sites, neg_sites):
    """Builds a response + features DataFrame by pulling pos_sites/neg_sites rows out of
    dat_feats. pos_sites/neg_sites: lists of dat_feats.index values - the positive (X1) and
    negative (X0) training sites.
    """
    overlap = set(pos_sites) & set(neg_sites)
    if overlap:
        raise ValueError(f"Sites cannot be in both pos_sites and neg_sites: {sorted(overlap)[:5]}")

    missing_pos = set(pos_sites) - set(dat_feats.index)
    if missing_pos:
        warnings.warn(f"{len(missing_pos)} pos_sites are not present in dat_feats and will be "
                      f"dropped: {', '.join(sorted(missing_pos)[:5])}")
        pos_sites = [s for s in pos_sites if s not in missing_pos]

    missing_neg = set(neg_sites) - set(dat_feats.index)
    if missing_neg:
        warnings.warn(f"{len(missing_neg)} neg_sites are not present in dat_feats and will be "
                      f"dropped: {', '.join(sorted(missing_neg)[:5])}")
        neg_sites = [s for s in neg_sites if s not in missing_neg]

    sites = list(pos_sites) + list(neg_sites)
    response = ["X1"] * len(pos_sites) + ["X0"] * len(neg_sites)
    train_dat = dat_feats.loc[sites].copy()
    train_dat.insert(0, "response", response)
    return train_dat


##### Full training pipeline #####

def run_training_pipeline(dat_feats, train_sites, num_reps=1, k=5, custom_splits=None,
                          split_seed=123, train_seed=80, return_all_preds_lst=False):
    """dat_feats: DataFrame of features for all sites (index = "<uniprot>_<site>"). Final
      predictions are made for every row of this DataFrame.
    train_sites: dict(pos=[...], neg=[...]) of dat_feats.index values, giving the positive
      (X1) and negative (X0) training sites. See load_default_train_sites() for the default.
    num_reps: number of repeated protein-stratified k-fold train-test splits to generate.
      Ignored if custom_splits is supplied
    k: number of folds per repeat
    custom_splits: optional pre-built train-test splits (as produced by
      make_folds_protstrat() / make_repeated_protstrat_splits()), used instead of
      auto-generating splits from num_reps/k
    split_seed: base seed used when auto-generating splits
    train_seed: base seed used for model training
    return_all_preds_lst: if True, also return each split's model's individual predictions
      on dat_feats (all_preds_df is already their per-site median across splits)

    Returns a dict (as produced by train_and_make_preds_logreg()):
      auc_df        - AUC of every train-test split, on its own held-out test set
      test_preds_df - the underlying test-set predictions behind auc_df
      all_preds_df  - median predicted score for every site in dat_feats, across all splits
      models        - the trained model for each split
      all_preds_lst - only present if return_all_preds_lst = True
    """
    train_dat = build_training_data(dat_feats, train_sites["pos"], train_sites["neg"])

    if custom_splits is None:
        train_test_splits = make_repeated_protstrat_splits(train_dat, num_reps=num_reps, k=k, seed=split_seed)
    else:
        train_test_splits = custom_splits

    return train_and_make_preds_logreg(
        train_test_splits, dat_feats, seed=train_seed, return_all_preds_lst=return_all_preds_lst
    )
